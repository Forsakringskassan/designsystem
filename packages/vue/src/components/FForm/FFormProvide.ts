import { Reference } from "@fkui/logic";
import { FormErrorList } from "../../types";

/**
 * @public
 */
export const setRef = Symbol("setRef");

/**
 * @public
 */
export const getRef = Symbol("getRef");

/**
 * @public
 */
export const setIsOpen = Symbol("setIsOpen");

/**
 * Methods provided by FForm.
 *
 * @public
 */
export interface FFormProvider {
    [setRef](id: string, data: FormErrorList): void;
    [getRef](id: string): Reference<FormErrorList>;
}

/**
 * @public
 */
export function createFFormProvideOptions(vm: {
    components: Record<string, Reference<FormErrorList>>;
}): FFormProvider {
    const { components } = vm;
    return {
        [setRef](id: string, data: FormErrorList): void {
            components[id] = new Reference(data);
        },
        [getRef](id: string): Reference<FormErrorList> {
            return components[id];
        },
    };
}
