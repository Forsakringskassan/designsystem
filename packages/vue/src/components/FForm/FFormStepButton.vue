<template>
    <button
        :id="id"
        class="button button--margin-bottom-0 button--large"
        type="button"
        v-bind="$attrs"
        :aria-expanded="isOpen ? 'true' : 'false'"
        data-form-step-button
    >
        <template v-if="isOpen">
            <f-icon name="success" class="button__icon" />
            {{ $t("fkui.form-step-button.close", "OK") }}
        </template>
        <template v-else>
            <f-icon name="pen" class="button__icon" />
            {{
                isAnyFieldTouched
                    ? $t("fkui.form-step-button.edit", "Ã„ndra")
                    : $t("fkui.form-step-button.open", "Fyll i")
            }}
        </template>
        <span class="sr-only">{{ additionalScreenreaderText }}</span>
    </button>
</template>

<script lang="ts">
import { defineComponent } from "vue";
import { ElementIdService } from "@fkui/logic";
import { FIcon } from "../FIcon";
import { TranslationMixin } from "../../plugins";

/**
 * @deprecated `<f-form-step-button>` is deprecated.
 * See migration guide: https://forsakringskassan.github.io/latest/components/fform.html
 */
export default defineComponent({
    name: "FFormStepButton",
    components: { FIcon },
    mixins: [TranslationMixin],
    inheritAttrs: false,
    model: {
        prop: "sr-only",
    },
    props: {
        /**
         * If the id is not set a random value will be generated.
         */
        id: {
            type: String,
            required: false,
            default: () => ElementIdService.generateElementId(),
        },
        isOpen: Boolean,
        isAnyFieldTouched: Boolean,
        additionalScreenreaderText: {
            type: String,
            required: false,
            default: "",
        },
    },
});
</script>
