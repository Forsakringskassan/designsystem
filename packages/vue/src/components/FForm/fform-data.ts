import { Reference } from "@fkui/logic";
import { FormErrorList } from "../../types";

/**
 * @public
 */
export interface FFormData {
    errorMessageSlotClass: string[];
    components: Record<string, Reference<FormErrorList>>;
}
