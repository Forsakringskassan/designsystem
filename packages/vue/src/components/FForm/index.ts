export { default as FForm } from "./FForm.vue";
export { default as FFormStep } from "./FFormStep.vue";
export {
    type FFormProvider,
    setRef,
    getRef,
    setIsOpen,
    createFFormProvideOptions,
} from "./FFormProvide";
export {
    type ComponentReferences,
    type ComponentValueTypes,
    cleanUpElements,
    isFormStepReference,
    sortComponentsWithErrorsOnDOMOrder,
} from "./FormUtils";
export { type FFormData } from "./fform-data";
