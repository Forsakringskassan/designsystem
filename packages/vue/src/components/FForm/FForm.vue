<template>
    <!-- [html-validate-disable-next wcag/h32: FForm does not have a default button and leaves the responsiblity to the application using FForm.] -->
    <form :id="id" class="form" novalidate v-bind="attrs" @component-validity="onComponentValidity">
        <f-message-box v-if="displayError && isSubmitted && hasError" tabindex="-1" type="error">
            <!--
@slot Slot for displaying error description.

After this slot a list of invalid elements is listed.
When an item is clicked it will scroll to that invalid element.

Slot content is available through `v-slot="{ <propertyName> }"`, e.g.
`v-slot="{ slotClass }"`.

The following properties are available:

* `slotClass: string[];` CSS classes to use for the error message heading element. Use with `:class="slotClass"`.
      -->
            <slot name="error-message" v-bind="{ slotClass: errorMessageSlotClass }"></slot>
            <nav>
                <f-error-list :items="componentsWithErrors"></f-error-list>
            </nav>
        </f-message-box>
        <!-- @slot Slot for displaying content, e.g. input elements and form steps. -->
        <slot name="default"></slot>
    </form>
</template>

<script lang="ts">
import { ElementIdService, focus, Reference, ValidationService } from "@fkui/logic";
import { defineComponent, type PropType } from "vue";
import { ComponentValidityEvent, FormErrorList, FormStep } from "../../types";
import { FMessageBox } from "../FMessageBox";
import { FErrorList } from "../FErrorList";
import { type FFormProvider, createFFormProvideOptions } from "./FFormProvide";
import { cleanUpElements, isFormStepReference, sortComponentsWithErrorsOnDOMOrder } from "./FormUtils";

interface FFormData {
    errorMessageSlotClass: string[];
    components: Record<string, Reference<FormErrorList | FormStep>>;
}

/**
 * @deprecated `<f-form>` is deprecated.
 * See migration guide: https://forsakringskassan.github.io/latest/components/fform.html
 */
export default defineComponent({
    name: "FForm",
    components: { FMessageBox, FErrorList },
    provide(): FFormProvider {
        return createFFormProvideOptions(this);
    },
    inheritAttrs: false,
    props: {
        /**
         * The id for the form id attribute.
         * If the prop is not set a random value will be generated.
         */
        id: {
            type: String,
            required: false,
            default: () => ElementIdService.generateElementId(),
        },
        /**
         * If the error-message slot and the invalid elements list
         * should be displayed.
         */
        displayError: {
            type: Boolean,
            default: true,
        },
        /**
         * If invalid fields at submit this tells how the scroll to the error
         * field should occur. The default, "center", centres the error information
         * vertically. Allowed value is also "top" which scrolls such that the error
         * informations top is at the top of the viewport.
         */
        errorScroll: {
            type: String as PropType<"center" | "top">,
            default: "center",
            validator(value: string) {
                return ["center", "top"].includes(value);
            },
        },
    },
    data(): FFormData {
        return {
            errorMessageSlotClass: ["message-box__heading"],
            components: {},
        };
    },
    computed: {
        attrs(): Record<string, unknown> {
            type SubmitHandler = (e: SubmitEvent) => void;
            let onSubmit: SubmitHandler | SubmitHandler[];
            // ? Merge submit listeners
            if (this.$attrs.onSubmit) {
                const originalSubmit = this.$attrs.onSubmit as SubmitHandler | SubmitHandler[];

                onSubmit = async (event: SubmitEvent) => {
                    const isValid = await this.onSubmit(event);
                    if (isValid) {
                        if (Array.isArray(originalSubmit)) {
                            originalSubmit.forEach((submitFunction) => submitFunction(event));
                        } else {
                            originalSubmit(event);
                        }
                    }
                };
            } else {
                onSubmit = this.onSubmit;
            }
            const attrs: Record<string, unknown> = { ...this.$attrs, id: this.id, onSubmit };
            return attrs;
        },
        numberOfTimesSubmitted(): number {
            const components: Array<Reference<FormErrorList | FormStep>> = Object.values(this.components);
            return Math.max(0, ...components.map((it) => it.ref.numberOfTimesSubmitted));
        },
        isSubmitted(): boolean {
            return this.numberOfTimesSubmitted > 0;
        },
        hasError(): boolean {
            if (Object.values(this.components).length === 0) {
                return false;
            } else {
                const components: Array<Reference<FormErrorList | FormStep>> = Object.values(this.components);
                return components.some((component) => component.ref.isValid === false);
            }
        },
        componentsWithErrors(): Array<FormErrorList | FormStep> {
            const sortedComponents = sortComponentsWithErrorsOnDOMOrder(this.components) as Array<
                Reference<FormErrorList | FormStep>
            >;

            return sortedComponents.map((c) => c.ref);
        },
        needsErrorScroll(): boolean {
            return this.errorScroll !== "center";
        },
    },
    methods: {
        async focusDomElement(): Promise<void> {
            await this.$nextTick();
            const messageBoxError = this.$el.querySelector(".message-box--error");
            if (messageBoxError) {
                focus(messageBoxError, { scrollToTop: this.needsErrorScroll });
            }
        },
        async focusFallbackElement(): Promise<void> {
            await this.$nextTick();
            const element = this.$el.querySelector(".form-step__header");
            const invalidElement = this.$el.querySelector(":invalid");
            if (element) {
                focus(element, { scrollToTop: this.needsErrorScroll });
            } else if (invalidElement) {
                focus(invalidElement, { force: true, scrollToTop: this.needsErrorScroll });
            }
        },
        async onSubmit(event: Event): Promise<boolean> {
            event.preventDefault();

            const components: Array<Reference<FormErrorList | FormStep>> = Object.values(this.components);
            components.forEach((component) => {
                component.ref.numberOfTimesSubmitted++;

                if (!isFormStepReference(component)) {
                    ValidationService.setTouched(component.ref.id);
                    ValidationService.setSubmitted(component.ref.id);
                }
            });

            await cleanUpElements(this);
            await ValidationService.validateAllElements(this.id);

            if (this.displayError) {
                this.focusDomElement();
            } else {
                // ! TODO: Focus on first invalid FormStep error list.
                this.focusFallbackElement();
            }

            return this.hasError === false;
        },
        async onComponentValidity(event: CustomEvent<ComponentValidityEvent>): Promise<void> {
            const reference = new Reference<FormErrorList | FormStep>({
                id: event.detail.elementId,
                focusElementId: event.detail.focusElementId,
                title: event.detail.errorMessage,
                isValid: event.detail.isValid,
                numberOfTimesSubmitted: this.numberOfTimesSubmitted,
            });

            this.components[event.detail.elementId] = reference;

            // required for components not within formsteps
            await cleanUpElements(this);
        },
    },
});
</script>
