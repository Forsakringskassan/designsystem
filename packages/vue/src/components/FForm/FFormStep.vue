<template>
    <f-validation-group
        :id="id"
        v-bind="attrs"
        :key="validationGroupKey"
        v-model="groupValidity"
        class="form-step"
        :stop-propagation="true"
    >
        <div class="form-step__header" tabindex="-1" role="group">
            <!-- @slot
Slot for header content, e.g. title.

Slot content is available through `v-slot="{ <propertyName> }"`, e.g.
`v-slot="{ slotClass }"`.

The following properties are available:

* `slotClass: string[];` CSS classes to use for the heading element. Use with `:class="slotClass"`.
* `+` all properties that are available in the default slot.
      -->
            <slot name="header" v-bind="{ slotClass: headerSlotClass, ...self.ref, toggleIsOpen }"></slot>
            <f-icon v-if="self.ref.isValid" class="form-step__check" name="success"></f-icon>
        </div>
        <div v-if="displayErrorList" :id="stepErrorId" class="form-step__error" tabindex="-1" role="group">
            <nav>
                <f-error-list :bullets="true" :items="componentsWithError" :before-navigate="beforeNavigateToError">
                    <template #title>
                        <!--
                             @slot Slot for displaying error description.

                             After this slot a list of invalid elements is listed.
                             When an item is clicked it will scroll to and focus that invalid element.
                        -->
                        <slot name="error-message">
                            {{ defaultErrorMessage }}
                        </slot>
                    </template>
                </f-error-list>
            </nav>
        </div>

        <div @component-validity="onComponentValidity">
            <!-- @slot
Slot for displaying content, e.g. descriptions and input elements.

Slot content is available through `v-slot="{ <propertyName> }"`, e.g.
`v-slot="{ isOpen, isSubmitted }"`.

The following properties are available:

* `isOpen: boolean;`
* `isSubmitted: boolean;`
* `isAnyFieldTouched: boolean;`
* `isValid: boolean;`
* `id: string;`
* `title: string;`
* `toggleIsOpen: () => void;`
    -->
            <slot name="default" v-bind="{ ...self.ref, isSubmitted, toggleIsOpen }"></slot>
            <span v-if="hasArrow" class="form-step__arrow"></span>
        </div>
    </f-validation-group>
</template>

<script lang="ts">
import { defineComponent, inject } from "vue";
import { ElementIdService, ValidationService, focus, Reference } from "@fkui/logic";
import { FormStep, ErrorItem, GroupValidityEvent, ComponentValidityEvent } from "../../types";
import { FIcon } from "../FIcon";
import { FErrorList } from "../FErrorList";
import { FValidationGroup } from "../FValidationGroup";
import { TranslationMixin } from "../../plugins";
import { setRef, getRef, setIsOpen } from "./FFormProvide";

/**
 * @deprecated `<f-form-step>` is deprecated.
 * See migration guide: https://forsakringskassan.github.io/latest/components/fform.html
 */
export default defineComponent({
    name: "FFormStep",
    components: { FIcon, FValidationGroup, FErrorList },
    mixins: [TranslationMixin],
    inheritAttrs: false,
    props: {
        /**
         * The id for the root div id attribute.
         * If the prop is not set a random value will be generated.
         */
        id: {
            type: String,
            required: true,
            default: () => ElementIdService.generateElementId(),
        },
        /**
         * If the form step should have a bottom arrow.
         * It is recommended to remove the arrow for the last step with `:has-arrow="false"`.
         */
        hasArrow: {
            type: Boolean,
            default: true,
        },
        /**
         * If the form step is the last step.
         * If set to true the form step will have a bigger bottom margin.
         */
        isLastStep: {
            type: Boolean,
            default: false,
        },
        /**
         * If error links should be disabled, `:disable-error-links="true"`..
         * If `true`, errors will be displayed as text instead.
         */
        disableErrorLinks: {
            type: Boolean,
            default: false,
        },
    },
    setup() {
        return {
            setRef: inject(setRef) as (id: string, data: FormStep | FormStep) => void,
            getRef: inject(getRef) as (id: string) => Reference<FormStep | FormStep>,
            setIsOpen: inject(setIsOpen) as (id: string, isOpen: boolean) => void,
        };
    },
    data() {
        return {
            self: new Reference<FormStep>({
                isOpen: false,
                numberOfTimesSubmitted: 0,
                isAnyFieldTouched: false,
                isValid: true,
                id: this.id,
                focusElementId: `${this.id}-error`,
                title: "",
            }),
            hasBeenValid: false,
            isOpenedClosed: false,
            headerSlotClass: ["form-step__title"],
            stepErrorId: `${this.id}-error`,
            groupValidity: {
                isValid: false,
                componentsWithError: [],
                componentCount: 0,
            } as GroupValidityEvent,
        };
    },
    computed: {
        attrs(): Record<string, string | Record<string, unknown>> {
            return {
                ...this.$attrs,
                class: {
                    "form-step--complete": this.self.ref.isValid,
                    "form-step--last-step": this.isLastStep,
                },
            };
        },
        defaultErrorMessage(): string {
            return this.$t("fkui.form-step.errorlist.title", "Oj, du har glömt att fylla i något. Gå till:");
        },
        // eslint-disable-next-line @typescript-eslint/no-explicit-any -- technical debt, not sure what is going on here
        injected(): any {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any -- technical debt, not sure what is going on here
            return this as any;
        },
        isSubmitted(): boolean {
            return this.self.ref.numberOfTimesSubmitted > 0;
        },
        componentsWithError(): ErrorItem[] {
            return this.groupValidity.componentsWithError.map((c: ComponentValidityEvent) => ({
                id: this.disableErrorLinks ? undefined : c.elementId,
                focusElementId: c.focusElementId,
                title: c.errorMessage,
            }));
        },
        hasError(): boolean {
            return this.groupValidity.componentsWithError.length > 0;
        },
        displayErrorList(): boolean {
            return (this.isOpenedClosed || this.hasBeenValid || this.isSubmitted) && this.hasError;
        },
        validationGroupKey(): string {
            return `${this.id}-group`;
        },
    },
    watch: {
        "self.ref.numberOfTimesSubmitted": {
            handler(): void {
                ValidationService.setSubmitted(this.$el);
                ValidationService.validateAllElements(this.id);
            },
        },
        groupValidity: {
            immediate: true,
            handler(): void {
                this.self.ref.isValid = this.groupValidity.isValid;

                if (this.hasBeenValid === false) {
                    this.hasBeenValid = this.groupValidity.isValid;
                }
            },
        },
    },
    async mounted(): Promise<void> {
        await this.$nextTick();
        this.loadSelf();
        this.injected.setRef(this.id, this.self.ref);
        this.loadSelf();
        this.updateSelfRefTitle();
    },
    beforeUpdate() {
        this.updateSelfRefTitle();
    },
    methods: {
        updateSelfRefTitle(): void {
            const [firstSlotClass] = this.headerSlotClass;
            const headerElement = this.$el.querySelector(`.${firstSlotClass}`);
            if (headerElement && headerElement.textContent) {
                this.self.ref.title = headerElement.textContent;
            }
        },
        loadSelf(): void {
            const ref = this.injected.getRef(this.id);
            Object.assign(this.self, ref);
        },
        async focusDomElement(): Promise<void> {
            await this.$nextTick();

            if (this.hasError && this.self.ref.isAnyFieldTouched) {
                const formStepError = this.$el.querySelector(".form-step__error");
                focus(formStepError);
            } else {
                const formStepHeader = this.$el.querySelector(".form-step__header");
                focus(formStepHeader);
            }
        },
        async beforeNavigateToError(): Promise<void> {
            this.injected.setIsOpen(this.id, true);
            await this.$nextTick();
        },
        /**
         * Toggle if open or closed.
         * If another form step is open then it will be closed.
         * @public
         */
        toggleIsOpen(): void {
            this.injected.setIsOpen(this.id, !this.self.ref.isOpen);
            // ? If any field is touched and form step was opened and closed.
            if (this.self.ref.isAnyFieldTouched && this.self.ref.isOpen === false) {
                if (this.isOpenedClosed === false) {
                    this.isOpenedClosed = true;
                }

                ValidationService.setTouched(this.$el);
                ValidationService.validateAllElements(this.id);
            }

            this.focusDomElement();
        },
        onComponentValidity(event: CustomEvent<ComponentValidityEvent>): void {
            const hasReceivedValidOrErrorEvent =
                event.detail.validityMode === "VALID" || event.detail.validityMode === "ERROR";

            if (hasReceivedValidOrErrorEvent && this.self.ref.isOpen) {
                this.self.ref.isAnyFieldTouched = true;
            }
        },
    },
});
</script>
