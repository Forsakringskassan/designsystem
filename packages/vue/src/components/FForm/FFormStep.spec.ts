import "html-validate/jest";
import "@fkui/test-utils/jest";
import { defineComponent } from "vue";
import { VueWrapper, mount, DOMWrapper } from "@vue/test-utils";
import flushPromises from "flush-promises";
import { ValidationService, Reference } from "@fkui/logic";
import { createPlaceholderInDocument } from "@fkui/test-utils/vue";
import { FormStep, FormErrorList } from "../../types";
import { ValidationPlugin } from "../../plugins";
import { FTextField } from "../FTextField";
import { FFieldset } from "../FFieldset";
import { FRadioField } from "../FRadioField";
import { FFormProvider, createFFormProvideOptions } from "./FFormProvide";
import FForm from "./FForm.vue";
import FFormStep from "./FFormStep.vue";

interface TestComponentData {
    components: Record<string, Reference<FormErrorList | FormStep>>;
}

const TestComponent = defineComponent({
    components: { FFormStep, FTextField },
    provide(): FFormProvider {
        return createFFormProvideOptions(this);
    },
    data(): TestComponentData {
        return {
            components: {},
        };
    },
    template: /* HTML */ `
        <f-form-step v-bind="$attrs">
            <template #header>
                <h2 class="form-step__title">example</h2>
            </template>
            <template #error-message>
                <div>error-message slot content</div>
            </template>
            <template #default="{ toggleIsOpen, isOpen }">
                <div v-show="isOpen">
                    <f-text-field
                        id="element1"
                        v-validation.required.maxLength="{ maxLength: { length: 40 } }"
                    />
                    <f-text-field
                        id="element2"
                        v-validation.required.maxLength="{ maxLength: { length: 40 } }"
                    />
                    <f-text-field id="element3" />
                </div>
                <button
                    id="button"
                    type="button"
                    @click="toggleIsOpen"
                ></button>
            </template>
        </f-form-step>
    `,
});

function createWrapper({
    props = {},
    slots = {},
    attrs = {},
    stubs = {},
} = {}): VueWrapper {
    return mount(TestComponent, {
        attrs: { ...attrs },
        props: { id: "test-id", titleText: "example", ...props },
        slots: { ...slots },
        attachTo: createPlaceholderInDocument(),
        global: {
            plugins: [ValidationPlugin],
            stubs: {
                FLabel: true,
                FIcon: true,
                FErrorList: {
                    props: ["items"],
                    template: /* HTML */ ` <div>{{ items }}</div> `,
                },
                ...stubs,
            },
        },
    });
}

async function touchInputs(
    wrapper: VueWrapper,
    inputs: Array<DOMWrapper<HTMLInputElement>>,
): Promise<void> {
    for (const input of inputs) {
        await input.trigger("focus");
    }
    await flushPromises();

    for (const input of inputs) {
        await input.trigger("blur");
    }
    await flushPromises();
}

async function writeInputs(
    wrapper: VueWrapper,
    inputs: Array<DOMWrapper<HTMLInputElement>>,
    value: string = "foobar",
): Promise<void> {
    for (const input of inputs) {
        await input.trigger("focus");
    }
    await wrapper.vm.$nextTick();
    await flushPromises();

    for (const input of inputs) {
        await input.setValue(value);
    }
    await wrapper.vm.$nextTick();
    await flushPromises();

    for (const input of inputs) {
        await input.trigger("blur");
    }
    await wrapper.vm.$nextTick();
    await flushPromises();
}

afterEach(() => {
    jest.restoreAllMocks();
});

describe("snapshots", () => {
    it("should match snapshot when pristine", async () => {
        expect.assertions(1);
        const wrapper = createWrapper();
        await flushPromises();
        expect(wrapper.element).toMatchSnapshot();
    });
});

describe("errorlist", () => {
    const errorList = ".form-step__error";

    it("should not be displayed initially", async () => {
        expect.assertions(1);
        const wrapper = createWrapper();
        await flushPromises();
        const formStep = wrapper.getComponent(FFormStep);
        expect(formStep.find(errorList).exists()).toBeFalsy();
    });

    it("should not be displayed when form step is opened and closed without touching input fields", async () => {
        expect.assertions(1);
        const wrapper = createWrapper();
        await flushPromises();
        const formStep = wrapper.getComponent(FFormStep);
        const button = wrapper.get("#button");

        /* open step */
        await button.trigger("click");
        await flushPromises();

        /* close step */
        await button.trigger("click");
        await flushPromises();

        expect(formStep.find(errorList).exists()).toBeFalsy();
    });

    it("should not be displayed when form step is opened and closed with touched input fields", async () => {
        expect.assertions(1);
        const wrapper = createWrapper();
        await flushPromises();
        const formStep = wrapper.getComponent(FFormStep);
        const button = wrapper.get("#button");
        const inputs = wrapper.findAll("input");

        /* open step */
        await button.trigger("click");
        await flushPromises();

        /* touch input fields */
        await touchInputs(wrapper, inputs);

        /* close step, the errorlist should now be visible */
        await button.trigger("click");
        await flushPromises();

        const element = formStep.find(errorList);
        expect(element.exists()).toBeFalsy();
    });

    it("should be displayed when form step is opened and closed with touched input fields having invalid values", async () => {
        expect.assertions(2);
        const wrapper = createWrapper();
        await flushPromises();
        const formStep = wrapper.getComponent(FFormStep);
        const button = wrapper.get("#button");
        const inputs = wrapper.findAll("input");

        /* open step */
        await button.trigger("click");
        await flushPromises();

        /* type to long value in input fields */
        await writeInputs(
            wrapper,
            inputs,
            "longer than maxlength validator of this field",
        );

        /* close step, the errorlist should now be visible */
        await button.trigger("click");
        await flushPromises();

        const element = formStep.find(errorList);
        expect(element.exists()).toBeTruthy();
        expect(element.element).toMatchSnapshot();
    });

    it("should be displayed without links when disableErrorLinks is enabled", async () => {
        const wrapper = mount(TestComponent, {
            props: {
                id: "test-id",
                titleText: "example",
                disableErrorLinks: true,
            },
            attachTo: createPlaceholderInDocument(),
            global: {
                plugins: [ValidationPlugin],
                stubs: {
                    FLabel: true,
                    FIcon: true,
                },
            },
        });

        const errorListBulletSelector = ".error-list__bullet";

        await flushPromises();
        const formStep = wrapper.getComponent(FFormStep);

        const button = wrapper.get("#button");
        const inputs = wrapper.findAll("input");

        /* open step */
        await button.trigger("click");
        await flushPromises();

        /* type to long value in input fields */
        await writeInputs(
            wrapper,
            inputs,
            "longer than maxlength validator of this field",
        );

        /* close step, the errorlist should now be visible without links */
        await button.trigger("click");
        await flushPromises();
        const errors = formStep.find(errorList);
        expect(errors.isVisible()).toBeTruthy();
        expect(errors.find("a").exists()).toBeFalsy();

        const bullets = formStep
            .find(errorList)
            .findAll(errorListBulletSelector);
        expect(bullets[0].element).toBeTruthy();
        expect(bullets[1].element).toBeTruthy();

        expect(formStep.props().disableErrorLinks).toBeTruthy();
    });

    it.todo("should be displayed when form is submitted");
});

describe("focus", () => {
    it("should not show errorlist when closed with only `required` form errors and touched fields", async () => {
        expect.assertions(1);

        /* given: a form step with input fields */
        const wrapper = createWrapper();
        await flushPromises();

        const button = wrapper.get("#button");
        const inputs = wrapper.findAll("input");

        /* when: form step is opened and closed, touching form fields in the
         * progress */
        await button.trigger("click");
        await flushPromises();
        await touchInputs(wrapper, inputs);
        await button.trigger("click");
        await flushPromises();

        /* then: the error list should not appear */
        expect(wrapper.find(".form-step__error").exists()).toBeFalsy();
    });

    it("should focus on errorlist when closed with other than `required` form errors and touched fields", async () => {
        expect.assertions(1);

        /* given: a form step with input fields */
        const wrapper = createWrapper();
        await flushPromises();

        const button = wrapper.get("#button");
        const inputs = wrapper.findAll("input");

        /* when: form step is opened and closed, touching form fields in the
         * progress */
        await button.trigger("click");
        await flushPromises();

        /* type to long value in input fields */
        await writeInputs(
            wrapper,
            inputs,
            "longer than maxlength validator of this field",
        );

        await button.trigger("click");
        await flushPromises();

        /* then: the error list should receive focus */
        const errorlist = wrapper.get(".form-step__error");
        expect(errorlist.element).toHaveFocus();
    });

    it("should focus on header when step is closed with form errors but untouched", async () => {
        expect.assertions(1);

        /* given: a form step with input fields */
        const wrapper = createWrapper();
        await flushPromises();

        /* when: form step is opened and closed without touching any form
         * fields */
        const button = wrapper.get("#button");
        await button.trigger("click");
        await flushPromises();
        await button.trigger("click");
        await flushPromises();

        /* then: the heading should receive focus */
        const header = wrapper.get(".form-step__header");
        expect(header.element).toHaveFocus();
    });

    it("should focus on header when form is closed without errors", async () => {
        expect.assertions(1);

        /* given: a form step with input fields */
        const wrapper = createWrapper();
        await flushPromises();

        const button = wrapper.get("#button");
        const inputs = wrapper.findAll("input");

        /* when: form step is opened and closed, touching form fields in the
         * progress */
        await button.trigger("click");
        await flushPromises();
        await writeInputs(wrapper, inputs);
        await button.trigger("click");
        await flushPromises();

        /* then: the heading should receive focus */
        const header = wrapper.get(".form-step__header");
        expect(header.element).toHaveFocus();
    });
});

describe("submit", () => {
    it("should call ValidationService.setSubmitted and ValidateService.validateAllElements when self.ref.numberOfTimeSubmitted is incremented", async () => {
        const validateAllElements = jest.spyOn(
            ValidationService,
            "validateAllElements",
        );
        const setState = jest.spyOn(ValidationService, "setSubmitted");

        const wrapper = createWrapper();
        const formStep = wrapper.getComponent(FFormStep);
        formStep.vm.$data.self.ref.numberOfTimesSubmitted++;
        await flushPromises();

        expect(setState).toHaveBeenCalled();
        expect(validateAllElements).toHaveBeenCalledTimes(1);
        expect(validateAllElements).toHaveBeenCalledWith("test-id");
    });
});

describe("error list", () => {
    it("should remove errors when field not in DOM anymore (i.e. v-if=false)", async () => {
        const Component = defineComponent({
            name: "TestComponent",
            components: {
                FForm,
                FFormStep,
                FTextField,
                FFieldset,
                FRadioField,
            },
            data() {
                return {
                    showField: "yes",
                    myText: "",
                };
            },
            template: /* HTML */ `
                <f-form :displayError="false">
                    <f-form-step id="form-step">
                        <f-fieldset
                            id="radiogroup"
                            name="myRadios"
                            v-validation.required
                        >
                            <template #label> Show field </template>
                            <template #default>
                                <f-radio-field
                                    id="radioRemoveField"
                                    value="no"
                                    v-model="showField"
                                >
                                    No
                                </f-radio-field>
                                <f-radio-field
                                    id="radioAddField"
                                    value="yes"
                                    v-model="showField"
                                >
                                    Yes
                                </f-radio-field>
                                <f-text-field
                                    v-if="showField==='yes'"
                                    id="textfield"
                                    v-model="myText"
                                    v-validation.required
                                >
                                    Textfield
                                </f-text-field>
                            </template>
                        </f-fieldset>
                    </f-form-step>
                </f-form>
            `,
        });
        const wrapper = mount(Component, {
            attachTo: createPlaceholderInDocument(),
            global: {
                plugins: [ValidationPlugin],
            },
        });

        await wrapper.get("#textfield").trigger("blur");
        await wrapper.get("form").trigger("submit");
        await flushPromises();

        const ul = wrapper.get(".error-list ul");
        expect(ul.html()).toMatchInlineSnapshot(`
            <ul class="error-list__list error-list--list-style-none">
              <li class=""><a href="javascript:"><span class="error-list__bullet" aria-hidden="true"></span><span class="error-list__link">Textfield</span></a></li>
            </ul>
        `);

        await wrapper.get("#radioRemoveField").trigger("click");
        await flushPromises();

        expect(wrapper.find(".error-list").exists()).toBeFalsy();
    });
});

describe("html-validate", () => {
    it.each`
        html
        ${"<f-form-step>Default</f-form-step>"}
        ${"<f-form-step><template #header>Title</template></f-form-step>"}
        ${"<f-form-step><template #error-message>Error</template></f-form-step>"}
    `("$html should be valid", ({ html }) => {
        expect.assertions(1);

        expect(html).toHTMLValidate();
    });

    it.each`
        html
        ${"<f-form-step><template #tooltip>Tooltip</template></f-form-step>"}
    `("$html should be invalid", ({ html }) => {
        expect.assertions(3);
        let catchedError;

        try {
            expect(html).toHTMLValidate();
        } catch (error) {
            catchedError = error;
        } finally {
            expect(catchedError).toBeDefined();
            expect(catchedError).toMatchSnapshot();
        }
    });
});
