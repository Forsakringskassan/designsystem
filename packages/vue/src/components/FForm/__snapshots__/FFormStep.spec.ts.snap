// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errorlist should be displayed when form step is opened and closed with touched input fields having invalid values 1`] = `
<div
  class="form-step__error"
  id="test-id-error"
  role="group"
  tabindex="-1"
>
  <nav>
    <div
      before-navigate="function () { [native code] }"
      bullets="true"
    >
      [
  {
    "id": "element1",
    "focusElementId": "element1",
    "title": ""
  },
  {
    "id": "element2",
    "focusElementId": "element2",
    "title": ""
  }
]
    </div>
  </nav>
</div>
`;

exports[`html-validate <f-form-step><template #tooltip>Tooltip</template></f-form-step> should be invalid 1`] = `
[Error: Expected HTML to be valid but had the following errors:

  <f-form-step> component has no slot "tooltip" [vue/available-slots]]
`;

exports[`snapshots should match snapshot when pristine 1`] = `
<div
  class="form-step"
  id="test-id"
  titletext="example"
>
  
  <div
    class="form-step__header"
    role="group"
    tabindex="-1"
  >
    
    <h2
      class="form-step__title"
    >
      example
    </h2>
    
    <!--v-if-->
  </div>
  <!--v-if-->
  <div>
    
    <div
      style="display: none;"
    >
      <div
        class="text-field"
      >
        <div
          class="i-width-sm-12"
        >
          <f-label-stub
            class=""
            for="element1"
          />
        </div>
        <div
          class="text-field__input-wrapper i-width-sm-12"
        >
          
          
          <div
            class="text-field__icon-wrapper"
          >
            <input
              aria-invalid="false"
              class="text-field__input"
              data-required=""
              data-validation=""
              id="element1"
              required=""
              type="text"
            />
            <!--v-if-->
            <!--v-if-->
            <!--v-if-->
            <!--v-if-->
          </div>
          
          
        </div>
        <!--v-if-->
      </div>
      <div
        class="text-field"
      >
        <div
          class="i-width-sm-12"
        >
          <f-label-stub
            class=""
            for="element2"
          />
        </div>
        <div
          class="text-field__input-wrapper i-width-sm-12"
        >
          
          
          <div
            class="text-field__icon-wrapper"
          >
            <input
              aria-invalid="false"
              class="text-field__input"
              data-required=""
              data-validation=""
              id="element2"
              required=""
              type="text"
            />
            <!--v-if-->
            <!--v-if-->
            <!--v-if-->
            <!--v-if-->
          </div>
          
          
        </div>
        <!--v-if-->
      </div>
      <div
        class="text-field"
      >
        <div
          class="i-width-sm-12"
        >
          <f-label-stub
            class=""
            for="element3"
          />
        </div>
        <div
          class="text-field__input-wrapper i-width-sm-12"
        >
          
          
          <div
            class="text-field__icon-wrapper"
          >
            <input
              class="text-field__input"
              id="element3"
              type="text"
            />
            <!--v-if-->
            <!--v-if-->
            <!--v-if-->
            <!--v-if-->
          </div>
          
          
        </div>
        <!--v-if-->
      </div>
    </div>
    <button
      id="button"
      type="button"
    />
    
    <span
      class="form-step__arrow"
    />
  </div>
  
</div>
`;
